### ISSUE TITLE
Class cannot assign to itself when using stub file

### ISSUE THREAD

**Issue by RF-Tar-Railt** [microsoft/pyright#10917]
Consider the following files:
```
- foo.py
- foo.pyi
- bar.py
```
with `foo.py`:

```python
from bar import bar

class Foo:
    id: str

    def __init__(self) -> None:
        bar.foos[self.id] = self
```

`foo.pyi`:
```python
class Foo:
    id: str

    def __init__(self) -> None: ...
```

`bar.py`:
```python
from typing import TYPE_CHECKING

if TYPE_CHECKING:
    from foo import Foo

class Bar:
    foos: dict[str, "Foo"]

bar= Bar()
```

And in `foo.py`, it reports error (on 1.1.405, Python 3.11):
```python
# Pyright: Argument of type "Self@Foo" cannot be assigned to parameter "value" of type "Foo" in function "__setitem__"  
#         "Foo*" is not assignable to "Foo"
bar.foos[self.id] = self
```
Which should not report.

**Additionally, it will not report if we don't have the `foo.pyi`**

**Comment by erictraut** [comment]
Pyright is working correctly here, so I don't consider this a bug.

You're defining `Foo` twice in two different files — a ".py" and a ".pyi" file. These are two different classes. The fact that they happen to have the same name is incidental. 

It looks like you're using a stub file here in an incorrect manner. A stub file is intended to be a stand-in for a library module that's implemented in another language or is missing type information.
